ArduPlane/quadplane.cpp

```diff
@@ -2276,48 +2276,77 @@ bool QuadPlane::verify_vtol_takeoff(const AP_Mission::Mission_Command &cmd)
 }
 
 /*
-  check if a landing is complete
+  a landing detector based on change in altitude over a timeout
  */
-void QuadPlane::check_land_complete(void)
+bool QuadPlane::land_detector(uint32_t timeout_ms)
 {
-    if (poscontrol.state != QPOS_LAND_FINAL) {
-        // only apply to final landing phase
-        return;
-    }
     const uint32_t now = AP_HAL::millis();
     bool might_be_landed =  (landing_detect.lower_limit_start_ms != 0 &&
                              now - landing_detect.lower_limit_start_ms > 1000);
     if (!might_be_landed) {
         landing_detect.land_start_ms = 0;
-        return;
+        return false;
     }
     float height = inertial_nav.get_altitude()*0.01f;
     if (landing_detect.land_start_ms == 0) {
         landing_detect.land_start_ms = now;
         landing_detect.vpos_start_m = height;
     }
+
     // we only consider the vehicle landed when the motors have been
-    // at minimum for 5s and the vertical position estimate has not
-    // changed by more than 20cm for 4s
+    // at minimum for timeout_ms+1000 and the vertical position estimate has not
+    // changed by more than 20cm for timeout_ms
     if (fabsf(height - landing_detect.vpos_start_m) > 0.2) {
         // height has changed, call off landing detection
         landing_detect.land_start_ms = 0;
-        return;
+        return false;
     }
            
-    if ((now - landing_detect.land_start_ms) < 4000 ||
-        (now - landing_detect.lower_limit_start_ms) < 5000) {
+    if ((now - landing_detect.land_start_ms) < timeout_ms ||
+        (now - landing_detect.lower_limit_start_ms) < (timeout_ms+1000)) {
         // not landed yet
-        return;
+        return false;
     }
     landing_detect.land_start_ms = 0;
-    // motors have been at zero for 5s, and we have had less than 0.3m
-    // change in altitude for last 4s. We are landed.
-    plane.disarm_motors();
-    poscontrol.state = QPOS_LAND_COMPLETE;
-    gcs().send_text(MAV_SEVERITY_INFO,"Land complete");
-    // reload target airspeed which could have been modified by the mission
-    plane.aparm.airspeed_cruise_cm.load();
+    landing_detect.lower_limit_start_ms = 0;
```

Comment:
Why reset this? `should_relax` is managing this for you already, and the affect of resetting this is that `should_relax` will fail for the next second. This can mean that the vehicle was on the ground and comes out of being relaxed on the position controller and will try and navigate, which can lead to tip overs/prop strikes.